### ISSUE TITLE
Gradient for SVD Op? 

### ISSUE THREAD

**Issue by Sentient07** [Theano/Theano#5440]
I'd like to perform SVD on the weights in a DNN. I think it would be useful to have the Gradient for SVD. I will implement this and open a PR if there are no alternative/easier ways. (which I am not aware of). 
cc : @nouiz @lamblin

**Comment by nouiz** [comment]
I don't know a better way then implementing an SVD op. Make sure to check
the code this generate. Maybe you could build a graph for the gradient. If
that is the case, it will be faster to have something working on the GPU.

**Comment by Sentient07** [comment]
> Make sure to check the code this generate. Maybe you could build a graph for the gradient. If
that is the case, it will be faster to have something working on the GPU.

You mean, we could have a C code for this Op and implement it (and its gradient) on the GPU?

**Comment by nouiz** [comment]
There is 2 choices on how to implement it:

1) build a graph of existing ops. Like the elemwise grad that reuse the
Elemwise op in its grad.
2) Make a new op. This will request ideally, CPU an GPU c code.

If 1 is possible, it will make it quick to have something working on the
GPU.

Note, I have no idea what the grad of SVD look like.

**Comment by Sentient07** [comment]
Regarding the second option that you have suggested, the existing Op makes a call to the numpy's implementation. If I have to make a new Op, for GPU, i will implement this,  http://docs.nvidia.com/cuda/cusolver/#svd-example1 , but for CPU, would writing C code make a significant difference ? Personally, I think most of the users will be using GPU while training their network and emphasising more on the CPU Op wouldn't be ideal. Correct me if I am missing something :)

**Comment by nouiz** [comment]
I agree, having a CPU op with only python code and a GPU op with C code is
the most useful case.

But if you can build a graph, then you won't need new op for the gradient!
That will make the code even simpler.

**Comment by f0k** [comment]
> Note, I have no idea what the grad of SVD look like.

The Jacobian of the singular values is easy. If X = U S V is your SVD, then UV is the Jacobian of the singular values S wrt. X. The Jacobian of U and V is more difficult. See [Estimating the Jacobian of the Singular Value Decomposition: Theory and Applications](https://hal.inria.fr/inria-00072686/document). @Sentient07, it'd definitely be useful to have that grad(), even if only on CPU.

**Comment by Sentient07** [comment]
Thanks a lot @f0k ! It was that material I was referring to while I opened this issue. 
@nouiz For clarification, when I apply SVD to a matrix and compile a function, it doesn't compile and I couldn't use the output of SVD for other purpose (like an input to an activation function). If that is what you mean by build a graph, then no. Also correct me if I am wrong, isn't every Op inside `nlinalg` uses numpy's Op and with that we can't construct a graph? (even though not everything would require a gradient or a graph, I am asking this out of curiosity)

**Comment by f0k** [comment]
Frédéric was referring to two options for `grad()`: a) Have it return the gradient expression in terms of existing basic Theano Ops, or b) write a specific SVDGrad Op that implements the gradient in Python/numpy or C. For example, for the Square Op, `grad()` can just return 2*input*grad_out. This is what was meant by building a graph, and it allows the gradient to be computed on GPU if the Ops involved support this anyway.

> For clarification, when I apply SVD to a matrix and compile a function, it doesn't compile and I couldn't use the output of SVD for other purpose (like an input to an activation function).

This should be possible already. You just can't compute the gradient wrt. to the SVD input.

**Comment by guoxuesong** [comment]
Before you complete gradient for SVD Op, maybe someone need this NormNucOp:

```
import numpy as np
import theano
from theano import gof
from theano import tensor as T
import theano.tests.unittest_tools

# Sum of the singular values.
#
class NormNucOp(gof.Op):
    _numop = staticmethod(np.linalg.svd)
    op = lambda x:np.asarray(self._numop(x, False, False).sum())

    def make_node(self, x):
        x = theano.tensor.as_tensor_variable(x)
        assert x.ndim == 2, "The input of svd function should be a matrix."
        s = theano.tensor.scalar(dtype=x.dtype)
        return gof.Apply(self, [x], [s])

    def perform(self, node, inputs, outputs):
        (x,) = inputs
        assert x.ndim == 2, "The input of svd function should be a matrix."
        s, = outputs
        s[0] = np.asarray(self._numop(x, False, False).sum())

    def grad(self, inputs, g_outputs):
        (x,) = inputs
        (gz,) = g_outputs
        assert x.ndim == 2, "The input of svd function should be a matrix."
        u, s, vt = theano.tensor.nlinalg.svd(x,False,True)
        d = u.dot(vt)
        return [gz*d]

    def infer_shape(self, node, shapes):
        return [()]

    def test_grad(self):
        theano.tests.unittest_tools.verify_grad(self, [np.random.normal(0,1,(5, 7,))])



def norm_nuc(a):
    return NormNucOp()(a)

#op = NormNucOp()
#op.test_grad()

```
